components/script/dom/shadowroot.rs

```diff
@@ -78,13 +79,25 @@ impl ShadowRoot {
     }
 
     pub fn detach(&self) {
-        self.host.set(None);
+        self.document.unregister_shadow_root(&self);
         let node = self.upcast::<Node>();
         node.set_containing_shadow_root(None);
-        node.set_flag(NodeFlags::IS_CONNECTED, false);
         for child in node.traverse_preorder(ShadowIncluding::No) {
-            child.set_flag(NodeFlags::IS_CONNECTED, false);
+            if node.RemoveChild(&child).is_err() {
+                warn!("Could not remove shadow root child");
+            }
+        }
+        if self
+            .host
+            .get()
+            .unwrap()
+            .upcast::<Node>()
+            .RemoveChild(&node)
+            .is_err()
+        {
+            warn!("Could not detach shadow root");
```

Comment:
So, this is not enough, you need to do the flags cleanup, and this will always error. Would it make sense to put [these two loops](https://github.com/servo/servo/blob/f4ff9e2eeb254a81f7b4ec9a43fe01fed2777b1a/components/script/dom/node.rs#L342) on their own function, then calling that from here and un-setting the host?

You may not need the `RemoveChild` then.